src/utils/size-observer.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ContentSizeObserverCallbackDef;
+
+/** @const {!WeakMap<!Window, !ResizeObserver>} */
+const observers = new WeakMap();
+
+/** @const {!WeakMap<!Element, !ContentSizeObserverCallbackDef>} */
+const targetObserverListMap = new WeakMap();
+
+/** @const {!WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+const targetSizeListMap = new WeakMap();
+
+/**
+ * @param {!Element} element
+ * @param {!ContentSizeObserverCallbackDef} callback
+ */
+export function observeContentSize(element, callback) {
+  const win = element.ownerDocument.defaultView;
+  if (!win) {
+    return;
+  }
+  let callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    callbacks = [];
+    targetObserverListMap.set(element, callbacks);
+    getObserver(win).observe(element);
+  }
+  if (pushIfNotExist(callbacks, callback)) {
+    const size = targetSizeListMap.get(element);
+    if (size) {
+      setTimeout(() => callCallbackNoInline(callback, size));
+    }
+  }
+}
+
+/**
+ * @param {!Element} element
+ * @param {!ObserverCallbackDef} callback
+ */
+export function unobserveContentSize(element, callback) {
+  const callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    return;
+  }
+  removeItem(callbacks, callback);
+  if (callbacks.length == 0) {
+    targetObserverListMap.delete(element);
+    targetSizeListMap.delete(element);
+    const win = element.ownerDocument.defaultView;
+    if (win) {
```

Comment:
nit: this is pretty edgecasey so I'm not concerned about impact, but technically if an element became detached before unobserved then this logic would never unobserve the element.

Reply:
Similar to the response above. Detachment is not an issue here. It's only when the whole iframe is destroyed. There's this sporadic time where all kinds of errors can happen due to messy deallocation. This only affects amount of errors we get that we can't do anything about anyway.

Reply:
Why isn't detachment an issue? Wouldn't it make the element windowless? (defaultView-less)
And either way, this sounds fine then!

Reply:
Detachment does not make element windowless. Per `element.ownerDocument.defaultView`, the window is strictly document's property, not element's. So only document detachment is an issue.